interfaces/builtin/wayland.go

```diff
@@ -19,32 +19,127 @@
   wayland:
     allow-installation:
       slot-snap-type:
+        - app
         - core
+    deny-auto-connection:
+      on-classic: false
 `
 
-const waylandConnectedPlugAppArmor = `
-# Description: Can access compositors supporting the wayland protocol
+const waylandPermanentSlotAppArmor = `
+# Description: Allow operating as a Wayland display server. This gives privileged access
+# to the system.
+
+# needed since Wayland is a display server and needs to configure tty devices
+capability sys_tty_config,
+/dev/tty[0-9]* rw,
+
+# Create the Wayland socket and lock file
+/run/user/[0-9]*/wayland-[0-9]* rw,
+
+# Allow write access to create XDG_RUNTIME_DIR (until lp:1656340 is fixed)
+/run/user/[0-9]*/ w,
+
+# Needed for mode setting via drmSetMaster() and drmDropMaster()
+capability sys_admin,
+
+# Weston probes this on start
+/sys/devices/pci**/boot_vga r,
+
+# NOTE: this allows reading and inserting all input events
+/dev/input/* rw,
 
-# Allow access to the wayland compsitor server socket
+# For using udev
+network netlink raw,
+/run/udev/data/c13:[0-9]* r,
+/run/udev/data/+input:input[0-9]* r,
+/run/udev/data/+platform:* r,
+`
+
+const waylandPermanentSlotSecComp = `
+# Description: Allow operating as a Wayland server. This gives privileged access
+# to the system.
+# Needed for server launch
+bind
+listen
+# Needed by server upon client connect
+accept4
+# for udev
+socket AF_NETLINK - NETLINK_KOBJECT_UEVENT
+`
+
+const waylandConnectedPlugAppArmor = `
+# Allow access to the Wayland compositor server socket
 owner /run/user/*/wayland-[0-9]* rw,
 
+# XWayland needs access to this
+/run/xwayland-shared-* rw,
+
 # Needed when using QT_QPA_PLATFORM=wayland-egl
 /etc/drirc r,
 `
 
+type waylandInterface struct{}
+
+func (iface *waylandInterface) Name() string {
+	return "wayland"
+}
+
+func (iface *waylandInterface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              waylandSummary,
+		ImplicitOnClassic:    true,
+		BaseDeclarationSlots: waylandBaseDeclarationSlots,
+	}
+}
+
+func (iface *waylandInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	spec.AddSnippet(slotAppLabelExpr(slot))
+	return nil
+}
+
+func (iface *waylandInterface) SecCompPermanentSlot(spec *seccomp.Specification, slot *snap.SlotInfo) error {
+	if !release.OnClassic {
+		spec.AddSnippet(waylandPermanentSlotSecComp)
+	}
+	return nil
+}
+
+func (iface *waylandInterface) AppArmorPermanentSlot(spec *apparmor.Specification, slot *snap.SlotInfo) error {
+	if !release.OnClassic {
+		spec.AddSnippet(waylandPermanentSlotAppArmor)
+	}
+	return nil
+}
+
+func (iface *waylandInterface) UDevPermanentSlot(spec *udev.Specification, slot *snap.SlotInfo) error {
+	spec.TagDevice(`KERNEL=="tty[0-9]*"`)
+	spec.TagDevice(`KERNEL=="mice"`)
+	spec.TagDevice(`KERNEL=="mouse[0-9]*"`)
+	spec.TagDevice(`KERNEL=="event[0-9]*"`)
+	spec.TagDevice(`KERNEL=="ts[0-9]*"`)
```

Comment:
These should be `if !release.OnClassic {` too I think.

Reply:
Including tty too?

Reply:
Yes. Note that this is the UDevPermanentSlot policy-- ie, what to tag for the slot provider so that the slot provider has these devices added to its device cgroup so it may function. When OnClassic, weston, mutter, etc are not running under snappy and are implicitly provided by the classic distro, and therefore there is no 'slot provider snap' where we need to tag these devices.

Reply:
Thanks for the explanation. Learning lots..